### PR TITLE
DEP: adjust requirements for extra dependency on typing-extensions

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

@hamogu @eerovaher

While reviewing #15685, I noticed that requirements on `typing_extensions` needed slight adjustements:
- we use it only for the `Self` symbol, which was added in version 4.0.0
- this symbol is in the standard library (`typing.Self`) in Python 3.11 and newer, so I made this requirement py310- only

I also found a note in a doctoring that doesn't seem relevant now that support for Python 3.8 was dropped.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I guess that the adjustment in minimal required version is technically a bug fix, so I'd suggest to backport this.

File: astropy/modeling/bounding_box.py [hunk 1]
```diff
@@ -17,12 +17,16 @@
...
     from typing import Any, Callable
 
-    from typing_extensions import Self
-
     from astropy.units import UnitBase
 
+    if sys.version_info >= (3, 11):
+        from typing import Self
+    else:
+        from typing_extensions import Self
```
Comment (eerovaher) [inline astropy/modeling/bounding_box.py:28]: This `if TYPE_CHECKING:` block was added in #15354. An early version of that pull request did include explicitly version dependent code, but during the review @nstarman pointed out that Ruff (specifically [UP035](https://docs.astral.sh/ruff/rules/deprecated-import/)) knows enough about `typing` and `typing_extensions` that it can automatically change the import when support for Python 3.10 is dropped. The change proposed here is not necessary.

File: astropy/units/quantity.py [hunk 2]
```diff
@@ -388,8 +388,7 @@ def __class_getitem__(cls, unit_shape_dtype):
 
         Notes
         -----
-        With Python 3.9+ or :mod:`typing_extensions`, |Quantity| types are also
-        static-type compatible.
+        |Quantity| types are also static-type compatible.
```
Comment (eerovaher) [inline astropy/units/quantity.py:391]: In the pull request @nstarman opened to drop support for Python 3.9 he also found some code and documentation that should or could have been updated when support for Python 3.8 was dropped. That pull request has not been merged yet. Perhaps it would be a good idea to open a separate pull request for removing the obsolete mentions of Python 3.9 like this.

File: pyproject.toml [hunk 3]
```diff
@@ -84,10 +84,10 @@ all = [
     "bottleneck",
     "ipython>=4.2",
     "pytest>=7.0",
-    "typing_extensions>=3.10.0.1",
     "fsspec[http]>=2023.4.0",
     "s3fs>=2023.4.0",
     "pre-commit",
+    "typing_extensions>=4.0.0 ; python_version < '3.11'",
```
Comment (eerovaher) [inline pyproject.toml:90]: It is good to try to keep track of why `typing_extensions` is needed, but it is also important to record the reason in commit messages.

File: astropy/units/quantity.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/units/quantity.py:391]: You must be referring to #15603. I notice this specific line isn't touched there. I'm happy to open another PR with just this (and possibly other similar changes if I find any), but I'm also fine if @nstarman wants to add it to his PR instead !

Comment (eerovaher) [reply astropy/units/quantity.py:391]: The problem with #15603 is that it seems to be stuck because the reference images for image comparison tests need to be regenerated. It would be better to split off things from there in new pull requests instead of adding more things to it that would then get stuck too.

Comment (nstarman) [reply astropy/units/quantity.py:391]: Agreed. Anything that can be paired off is good. Feel free to 🍒pick commits.

**Review by nstarman — APPROVED** [review]
Thanks @neutrinoceros! Looks good as is, but I have a small suggestion. Also @eerovaher's comment on comments is good." Lest we forget why `typing_extensions` is version-pinned

**Comment by neutrinoceros** [comment]
> Do we actually need typing_extensions as a direct dependency anyways?

it's currently only included as an extra dependency (part of the `all` target), which seems reasonable ?

File: astropy/units/quantity.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/units/quantity.py:391]: Ok, I'll move it to a separate branch, and will open a PR in a few days with maybe more of the same !

**Comment by neutrinoceros** [comment]
@eerovaher @nstarman I just removed the adjacent-but-not-strictly-related part of the patch and amended the commit message with more details.

**Comment by eerovaher** [comment]
If we don't list `typing_extensions` as a dependency at all then we don't have to worry about which Python versions it should be installed with, and then it's clear that we don't need any version dependent code inside `if TYPE_CHECKING:` blocks either.

**Comment by neutrinoceros** [comment]
Then I'll leave it to @nstarman to decide what to do here since he previously approved, but again, no harm done in closing this PR if we don't collectively think it's needed :)

**Comment by eerovaher** [comment]
I'm not suggesting to close this pull request without merging, I'm suggesting *remove* `typing_extensions` from the dependency list.

**Comment by nstarman** [comment]
I think if someone installs `Astropy[all]` they should get everything, including static type time stuff. Perhaps we should add an install option [type] that is included in all. Because with Protocols and API we can support interfaces without run-time dependencies. So IMO let's leave it `all` for now and follow up with a `type` install option

**Comment by neutrinoceros** [comment]
Fine by me @nstarman. I'm also happy to open another PR to implement your proposed `[type]` extra, if you'd like.

**Comment by nstarman** [comment]
> Fine by me @nstarman. I'm also happy to open another PR to implement your proposed `[type]` extra, if you'd like.

Much appreciated!

File: astropy/modeling/bounding_box.py (hunk 1 above)
Comment (nstarman) [reply astropy/modeling/bounding_box.py:28]: #minimum-effort.

**Comment by neutrinoceros** [comment]
done in https://github.com/astropy/astropy/pull/15689

**Comment by nstarman** [comment]
Merge this first or what's the order?

**Comment by eerovaher** [comment]
What is this pull request supposed to do when #15689 is merged?

**Comment by eerovaher** [comment]
I don't think the version check is necessary for preventing unwanted installations of `typing_extensions` now that #15689 is merged. If users install `astropy[typing]` they probably want `typing_extensions`. If they install `astropy[all]` they shouldn't be surprised if they end up with packages they are not actually using.

**Comment by neutrinoceros** [comment]
> If users install astropy[typing] they probably want typing_extensions.

I respectfully disagree here, but I won't insist. Again, I'm fine with this PR being closed if it's controversial.

**Comment by neutrinoceros** [comment]
I believe the only remaining failure is due to #15692, which I think can be resolved separately, so I'll start working on that now.

**Comment by pllim** [comment]
Will this fix https://github.com/astropy/astropy/issues/15819 ? If so, can we wrap this up for merge? Thanks!

**Comment by pllim** [comment]
Actually needs a rebase, and then I can slap "Extra CI" on this, to be sure.

**Comment by neutrinoceros** [comment]
> Will this fix https://github.com/astropy/astropy/issues/15819 ?

I don't think so, but I'll rebase now anyway !

**Comment by pllim** [comment]
Hmm link check is green here. 🤷

File: astropy/modeling/bounding_box.py (hunk 1 above)
Comment (pllim) [reply astropy/modeling/bounding_box.py:28]: Is this conversation resolved or not? Is this PR good to merge? @eerovaher ?

**Comment by nstarman** [comment]
My personal preference would be to just have `typing_extensions` as a standard dependency of Astropy and let `ruff` switch imports from `typing_extensions` to `typing` per our minimum Python version. `typing_extensions` is developed within the core Python org (https://github.com/python/typing_extensions), so this is a very safe approach. That said,  since we have not yet committed to having `typing_extensions` as a dependency, this PR is good to merge. Thanks @neutrinoceros!
